### ISSUE TITLE
Update DXVK for nvidia GPU

### ISSUE THREAD

**Issue by Kreyren** [PhoenicisOrg/scripts#982]
> You might need
> 
> > Run your game with VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/radeon_icd.x86_64.json for AMD 64 bit or with VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/nvidia_icd.json for nvidia.

Works now -> DXVK verb update is required

_Originally posted by @Kreyren in https://github.com/PhoenicisOrg/scripts/pull/965#issuecomment-499448221_

`VK_ICD_FILENAMES` variable has to be set depending on end-user system for games work.

**Comment by ImperatorS79** [comment]
We need a way to guess the main GPU on a system and a way to update the _environment variable of the QuickScript from the DXVK verb.

**Comment by Kreyren** [comment]
@ImperatorS79 Why guess? we can grab the vendor from linux environment

**Comment by ImperatorS79** [comment]
@plata How could I send the `QuickScript `to the dxvk verb so that I could update the `_environment `var ?

**Comment by ImperatorS79** [comment]
OK I have something that should work, we just have to change the prototype of `preInstall `and `postInstall `function:
```js
this._preInstall(this, wine, setupWizard);
```

**Comment by Kreyren** [comment]
> OK I have something that should work, we just have to change the prototype of `preInstall `and `postInstall `function:
> 
> ```js
> this._preInstall(this, wine, setupWizard);
> ```

Will you MR it?

**Comment by ImperatorS79** [comment]
It needs #956 first (which some changes in the GUI)

**Comment by plata** [comment]
I think we must query the user because there might be a multi GPU setup (will the system provide the correct GPU?).

**Comment by ImperatorS79** [comment]
@plata That's what I am doing locally, the fix is ready, but I want your opinion about the preInstall change  ^^.

**Comment by plata** [comment]
Hard to tell without actually seeing it.

**Comment by ImperatorS79** [comment]
Just changing the prototype to `this._preInstall(this, wine, setupWizard);` instead of `this._preInstall(wine, setupWizard);` in any file using it so that I can access script variables.

**Comment by plata** [comment]
I got that part. I'm interested in the part where you use the script variable.

**Comment by ImperatorS79** [comment]
I just get the `script._environment` variable (in my new shortcut pr), to append to it the `VK_ICD_FILENAMES variable`:

```js
Wine.prototype.DXVK = function (quickscript, dxvkVersion) {
    ...
    //Update environment to ensure the right GPU is loaded
    var gpuType = this.wizard().menu("Please select the type of your main GPU", ["AMD", "Intel", "Nvidia"]);
    var envJSON = {};
    if (typeof quickscript._environment !== 'undefined'){
        envJSON = JSON.parse(quickscript._environment);
    }
    if (gpuType.text == "Nvidia") {
        envJSON["VK_ICD_FILENAMES"] = "/usr/share/vulkan/icd.d/nvidia_icd.json";
    }
    else if (gpuType.text == "AMD") {
        if (this.architecture() == "amd64"){
            envJSON["VK_ICD_FILENAMES"] = "/usr/share/vulkan/icd.d/radeon_icd.x86_64.json";
        }
        else{
            envJSON["VK_ICD_FILENAMES"] = "/usr/share/vulkan/icd.d/radeon_icd.x86.json"; //To check
        }
    }
    quickscript._environment = JSON.stringify(envJSON);
    ...
```

**Comment by plata** [comment]
Won't this cause issues if you do not use a QuickScript?

**Comment by ImperatorS79** [comment]
All scripts inherit from QuickScript I think.

**Comment by plata** [comment]
No. It's convenient but not required.

**Comment by ImperatorS79** [comment]
Which one does not inherit ? I just have to add the same logic for them. 

Is there any reason why some script does not inherit from the same base ?

**Comment by plata** [comment]
For example: https://github.com/PhoenicisOrg/scripts/blob/master/Applications/Internet/Internet%20Explorer%207.0/Online/script.js. We want to have the possibility to write completely free scripts as well.

**Comment by ImperatorS79** [comment]
Okay so I have to totally rethink how this could work :thinking:.

**Comment by plata** [comment]
I fear so, yes. I also have no great idea right now.

**Comment by plata** [comment]
Except: Can you not simply return the environment from the function?

**Comment by ImperatorS79** [comment]
Yes, I could the json string or object! But I still need the modification of `.preInstall` ^^.

**Comment by plata** [comment]
Why? Once you are in the `preInstall`, you should be able to access a variable in the script. A bit like here: https://github.com/PhoenicisOrg/scripts/blob/master/Applications/Games/Rocksmith%202014/Steam/script.js

**Comment by ImperatorS79** [comment]
In my implementation, the environment variables are stored in a variable of `QuickScript `(and set with `.environment()` function of `QuickScript`). Since DXVK is executed in the preInstall, how could I modify that "private" variable without this ?

**Comment by plata** [comment]
Can't you just call `.environment()` after `.preInstall()`?

**Comment by ImperatorS79** [comment]
But how do I recover the variables outputted by DXVK inside the `.preInstall` ?

**Comment by plata** [comment]
Untested but can't you:
```Javascript
include("engines.wine.quick_script.steam_script");

var env = "";

var installerImplementation = {
    run: function () {
        new SteamScript()
            .preInstall(function (wine, wizard) {
                env =wine.DXVK();
            })
            .environment(env)
            .go();
    }
};
```

**Comment by ImperatorS79** [comment]
That seems really weird but worth trying.

**Comment by plata** [comment]
What if `environment()` could also receive a function, like:
```Javascript
.environment(getDxvkEnvironment)
```
?

**Comment by madoar** [comment]
That sounds like a beautiful solution!
